src/ripple/app/misc/NegativeUNLVote.h

```diff
@@ -0,0 +1,193 @@
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_APP_MISC_NEGATIVEUNLVOTE_H_INCLUDED
+#define RIPPLE_APP_MISC_NEGATIVEUNLVOTE_H_INCLUDED
+
+#include <ripple/beast/utility/Journal.h>
+#include <ripple/protocol/Protocol.h>
+#include <ripple/protocol/PublicKey.h>
+#include <ripple/protocol/UintTypes.h>
+
+#include <optional>
+
+namespace ripple {
+
+class Ledger;
+template <class Adaptor>
+class Validations;
+class RCLValidationsAdaptor;
+using RCLValidations = Validations<RCLValidationsAdaptor>;
+class SHAMap;
+namespace test {
+class NegativeUNLVoteInternal_test;
+class NegativeUNLVoteScoreTable_test;
+}  // namespace test
+
+/**
+ * Manager to create NegativeUNL votes.
+ */
+class NegativeUNLVote final
+{
+public:
+    /**
+     * A validator is considered unreliable if its validations is less than
+     * nUnlLowWaterMark in the last flag ledger period.
+     * An unreliable validator is a candidate to be disabled by the NegativeUNL
+     * protocol.
+     */
+    static constexpr size_t nUnlLowWaterMark = 256 * 0.5;
+    /**
+     * An unreliable validator must have more than nUnlHighWaterMark validations
+     * in the last flag ledger period to be re-enabled.
+     */
+    static constexpr size_t nUnlHighWaterMark = 256 * 0.8;
+    /**
+     * The minimum number of validations of the local node for it to
+     * participate in the voting.
+     */
+    static constexpr size_t nUnlMinLocalValsToVote = 256 * 0.9;
+    /**
+     * We don't want to disable new validators immediately after adding them.
+     * So we skip voting for disabling them for 2 flag ledgers.
+     */
+    static constexpr size_t newValidatorDisableSkip = 256 * 2;
+    /**
+     * We only want to put 25% of the UNL on the NegativeUNL.
+     */
+    static constexpr float nUnlMaxListed = 0.25;
+
+    /**
+     * Constructor
+     *
+     * @param myId the NodeID of the local node
+     * @param j log
+     */
+    NegativeUNLVote(NodeID const& myId, beast::Journal j);
+    ~NegativeUNLVote() = default;
+
+    /**
+     * Cast our local vote on the NegativeUNL candidates.
+     *
+     * @param prevLedger the parent ledger
+     * @param unlKeys the trusted master keys of validators in the UNL
+     * @param validations the validation message container
+     * @param initialSet the transactions set for adding ttUNL_MODIFY Tx if any
+     */
+    void
+    doVoting(
+        std::shared_ptr<Ledger const> const& prevLedger,
+        hash_set<PublicKey> const& unlKeys,
+        RCLValidations& validations,
+        std::shared_ptr<SHAMap> const& initialSet);
```

Comment:
This function can modify `validations`. That should be documented (see comment on `buildScoreTable`)

Reply:
Thanks, documented